### ISSUE TITLE
Stream hangs non-deterministically when using mapAsync

### ISSUE THREAD

**Issue by vlovgr** [functional-streams-for-scala/fs2#1297]
The following program:
```scala
import cats.effect.{ExitCode, IO, IOApp}
import cats.implicits._
import fs2.Stream

object Main extends IOApp {
  override def run(args: List[String]): IO[ExitCode] = {
    val stream =
      Stream
        .iterate(0)(_ + 1)
        .covary[IO]
        .mapAsync(25) { n =>
          if (n > 100) IO.raiseError(new RuntimeException)
          else IO(println(n))
        }

    stream.compile.drain.as(ExitCode.Success)
  }
}
```
when run somtimes hangs (never completes) indefinitely. It might take several runs/attempts for the hang to occur, but I'm always able to eventually trigger it locally.

**Comment by vlovgr** [comment]
Taking a look at `parEvalMap` (`mapAsync` is an alias), it makes use of `concurrently`. In this case, the error occurs in `this` Stream, so we have to interrupt the concurrently running Stream. Taking a look at `concurrently`, we're hitting `release` of the following bracket.

```scala
Stream.bracket(F.start(runR))(_ =>
  interrupt.complete(()).attempt >> // always interrupt `that`
    doneR.get.flatMap(F.fromEither) // always await `that` result
) >> this.interruptWhen(interrupt.get.attempt)
```

When this issue occurs, I've been able to verify that `interrupt` gets completed, but `doneR` never completes. I've also been able to verify that the `Fiber` running `runR` never completes, and removing the waiting on `doneR` makes me unable to reproduce this issue. The conclusion is that something with `interruptWhen` is not working properly.

Perhaps this is related to https://github.com/functional-streams-for-scala/fs2/issues/1300.

**Comment by SystemFw** [comment]
Thanks for reporting, I still haven't had any time to look into it unfortunately.

**Comment by vlovgr** [comment]
@SystemFw I've been able to identify and resolve the issue.

In `parEvalMap` we have the following code.

```scala
queue.dequeue.unNoneTerminate
  .evalMap(identity)
  .rethrow
  .concurrently {
    evalMap { o =>
      Deferred[F2, Either[Throwable, O2]].flatMap { value =>
        queue.enqueue1(Some(value.get)).as {
          Stream.eval(f(o).attempt).evalMap(value.complete)
        }
      }
    }.parJoin(maxConcurrent)
      .onFinalize(queue.enqueue1(None))
  }
```

When the dequeuing stream encounters an exception (from `rethrow`), we attempt to interrupt the concurrently running stream. However, there are two issues which can prevent interruption, because the queue might be full:

- `evalMap` never finishes because `queue.enqueue1` blocks forever,
- the finalizer never finishes, because `queue.enqueue1` blocks forever.

The solution is to use a `Deferred` and a finalizer on the dequeuing stream to signify that it has finished, and then cancel the blocking enqueues. I'll open a pull request with the fix shortly.

**Comment by SystemFw** [comment]
amazing! Thanks a lot, I'm swamped these days :)
